### PR TITLE
FEATURE: Added new middleware to rewrite the hash links and disabled SSViewer's rewriting by default

### LINKED ISSUE #8986: SSViewer anchor rewriting can be captured in cache blocks
## Affected Version

silverstripe/recipe-cms                                4.3.3

## Description

SSViewer's anchor/hash link rewritting can be captured in cache blocks in some instances due to how nested template rendering is performed. This should probably only be done after the cache block is loaded not when it is cached. As in some instances a link may have identifiable information for marketing purposes or other tracking information that is unique to the visitor. However with the rewriting being stored in the cache it's possible that the visitor with url parameters could re-generate the cache thus capturing their rewritten urls in the cache.

## Steps to Reproduce

Given the following template:

```
<% cached 'cacheName', $SomeCacheVariables %>
    <% loop $Elements %>
        $Me
    <% end_loop %>
<% end_if %>
```

And this example element template:

```html
<div class="element">
    <nav>
        <ul>
            <li><a href="#example-anchor-1">Example Anchor 1</a></li>
            <li><a href="#example-anchor-2">Example Anchor 2</a></li>
            <li><a href="#example-anchor-3">Example Anchor 3</a></li>
            <li><a href="#example-anchor-4">Example Anchor 4</a></li>
        </ul>
    </nav>
</div>
```

You could end up with something like the below in the actual rendered cache block, until the cache is regenerated.

```html
<div class="element">
    <nav>
        <ul>
            <li><a href="/my-page/?cid=us:media:192090&amp;dclid=ofnisdonfdisofnheu#example-anchor-1">Example Anchor 1</a></li>
            <li><a href="/my-page/?cid=us:media:192090&amp;dclid=ofnisdonfdisofnheu#example-anchor-2">Example Anchor 2</a></li>
            <li><a href="/my-page/?cid=us:media:192090&amp;dclid=ofnisdonfdisofnheu#example-anchor-3">Example Anchor 3</a></li>
            <li><a href="/my-page/?cid=us:media:192090&amp;dclid=ofnisdonfdisofnheu#example-anchor-4">Example Anchor 4</a></li>
        </ul>
    </nav>
</div>
```

### PR DESCRIPTION
**Pull request by UndefinedOffset**
While attempting to fix #8986 in pull #9044 I realized that the changes in that pull request don't work in all scenarios (see my comments there for my discoveries). This pull request introduces a new HTTP middleware that basically moves the hash rewriting from SSViewer which prevents the cache blocks from capturing the hash rewriting and exposing parameters in the url at the time of caching. If accepted I think the SSViewer hash rewriting should probably be fully deprecated and removed in 5 if this is (or something similar) is used in Silverstripe 5.

@emteknetnz Per our discussion here is the changes I was talking about, I am/Webbuilders Group is currently using this code on a couple sites with no known issues (so far) and it does seem to have corrected the issue for us.

### CODE REVIEW

**Comment by UndefinedOffset** [comment]
I should point out too that this technically solves a bug present in all supported versions of Silverstripe 4.x (likely 3.x as well), but I targeted 4 knowing that it's also a new way of approaching hash rewriting. I'll leave it to the core team to decide if it should be back ported to 4.5 and 4.6 😉 should this be accepted.

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (emteknetnz) [inline src/Control/Middleware/RewriteHashLinksMiddleware.php:42]:
```suggestion
            if (strpos($body, '<base') === false) {
               return;
            }
```
Then unindent the lines below

Comment (emteknetnz) [inline src/Control/Middleware/RewriteHashLinksMiddleware.php:42]: Do we also need to test for '<BASE' ?

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (emteknetnz) [inline src/Control/Middleware/RewriteHashLinksMiddleware.php:43]: Need to test for href='' singlequotes as well as href="" doublequotes

Comment (emteknetnz) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php:43]: `preg_replace("/^(\\/)+/"`

Probably cleaner to use str_replace here, I'm not entirely sure what characters are being replace due to the escaping

I'd also move `Convert::raw2att...` to the line above and assign it to a variable to increase readability

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (emteknetnz) [inline src/Control/Middleware/RewriteHashLinksMiddleware.php:38]:
```suggestion
        $mimeType = strtolower(trim(array_shift($contentType)));
```

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (emteknetnz) [inline src/Control/Middleware/RewriteHashLinksMiddleware.php:39]:
```suggestion
        if (!in_array($contentType, $this->config()->handled_mime_types)) {
          return;
        }  
```
And unindent the lines below

File: tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php
Comment (emteknetnz) [inline tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php:23]: Bunch of double blank lines in this file that should be reduces to a single blank line

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Mainly just stylistic changes, I'm probably not the best to judge if this approach in general is great idea or not.

File: src/View/SSViewer.php
Comment (UndefinedOffset) [reply src/View/SSViewer.php:115]: Ya it's just straight broken and it's not an easy fix it would require probably very very large changes to SSViewer to correct the issue. The middleware replaces the functionality in SSViewer so disabling it here prevents overlap and fixes the issue referenced.

**Comment by UndefinedOffset** [comment]
I'll get on the style changes next week ;) taking a little vacation but I did reply to your comment about SSViewer's functionality being turned off.

**Review by dhensby — CHANGES_REQUESTED** [review]
I think this is generally a good direction to go in - it's good to decouple this rewriting from the templating engine.

There are a few changes I've highlighted ~~and I'd also like to make sure that any existing tests that SSViewer had are still passing with this new approach too (ie: those tests probably need to be reimplemented)~~ Ah, looks like you did lift the tests from `SSViewerTest`.

File: src/View/SSViewer.php
Comment (UndefinedOffset) [reply src/View/SSViewer.php:115]: Ya I think that's a good idea. I thought I seen someway in the docs with I think it was the password expired middleware that allowed you to remove a single middlewhere which is probably why I didn't add a config for that.

**Comment by UndefinedOffset** [comment]
I'll get on the charges requested here Monday if I have time and I'll update when completed.

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (UndefinedOffset) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php:42]: I don't think this would be correct if there is no base tag there isn't really a need to do hash rewriting. In large part this code is a lift and move from SSViewer which is checking `!== false` https://github.com/silverstripe/silverstripe-framework/blob/4/src/View/SSViewer.php#L684-L694

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (UndefinedOffset) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php:43]: I like the idea however the one downside to this is that in the case of the home page `HTTPRequest::getURL(true)` returns `/ss_clients/silverstripe4/home` which would still cause the browser to navigate.

File: tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php
Comment (UndefinedOffset) [reply tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php:19]: Hmmm only problem is that especially with HTTP_HOST there's no other way to set that in the HTTPRequest. What do you think about backing up the values during test setup and tear down?

**Comment by kinglozzer** [comment]
One concern I have with this: what about anyone doing server-side rendering? They _might_ be using `Director::test()` (which I believe would trigger the middleware) if it’s possible to use it, but that isn’t always the case. E.g. one of our modules doesn’t: https://github.com/bigfork/silverstripe-fail-whale/blob/ba23f80ebc66cbdfcb1fbfc125bf2d02fc2083a8/src/Model/ErrorDocument.php#L324-L352 - the static HTML copies stored in `public/assets` would no longer have the hash rewriting performed on them as they won’t pass through the middleware if hit directly from Apache. 

We’ll definitely need a note added to the changelogs if we take this in 🙂

**Comment by UndefinedOffset** [comment]
@kinglozzer I agree this needs to be noted well in the change logs because of this, it's really the only option I could see to reliably fix #8986 short of re-writing probably a fairly large part of the template engine 😕 to avoid the caching issue.

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (dhensby) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php]: The core request object doesn't provide the actual request URL?!

File: tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php
Comment (dhensby) [reply tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php:19]: I'm pretty sure phpunit already backs up the globals. If it doesn't by default it can be enabled

Comment (UndefinedOffset) [reply tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php:19]: It looks like SapphireTest actually turns backup globals off "to avoid overriding the few globals SilverStripe relies on, like $lang for the i18n subsystem". I'll just implement something that backs up the host and request uri and resets it.

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (UndefinedOffset) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php]: I actually lifted that from SSViewer ;) seems to be replacing multiple back to back backslashes i.e `//` or `///` for example.

Comment (UndefinedOffset) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php]: I've implemented your other suggestions though here :)

File: tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php
Comment (dhensby) [inline tests/php/Control/Middleware/RewriteHashLinksMiddlewareTest.php:31]: you don't appear to be calling `parent::tearDown()`

**Review by dhensby — DISMISSED** [review]

File: src/Control/Middleware/RewriteHashLinksMiddleware.php
Comment (dhensby) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php]: from reviewing the HTTPRequest and the HTTPRequestBuilder I can't see how `getUrl()` will return `home` if `home` isn't the currently requested URL...

Comment (UndefinedOffset) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php]: It seems to be after the `$delegate($request)` so it must be one of the other `HTTPMiddleware` I'll just grab the value before the delegate call that should address the issue.

Comment (UndefinedOffset) [reply src/Control/Middleware/RewriteHashLinksMiddleware.php]: It is still not 100% accurate though for example visiting `http://example.com/?test` `getURL(true)` returns `?test=` which I think  would still result in a navigation but it's probably a rare case but something to consider anyways.

**Comment by dhensby** [comment]
Nice - I'd still prefer for us not to be relying on the super-globals, as a good PHP application we shouldn't be relying on state outside of the "app", but I'll leave it to others to decide if this should be a blocker or not

**Comment by UndefinedOffset** [comment]
I agree for sure it would be nice to lean on `HTTPRequest` fully, I used old code lifted from SSViewer in large part just modified for this application. I worry that the differences between what `HTTPRequest::getURL(true)` returns (i.e the mentioned `?test=` when visiting `http://example.com/?test` will still cause the browsers to navigate since it differs from the url the user visited.

**Comment by GuySartorelli** [comment]
I'm going to close this - I think it's fundamentally a bad approach for the problem. Rewriting something for the template layer should be handled by the template layer, not in a middleware.

**Pull request closed without merging**
